File: iosApp/iosApp/LoginViewModel.swift
//
//  LoginViewModel.swift
//  iosApp
//
//  Created by Iqbal Fauzi on 15/08/25.
//
import Shared

@MainActor
final class LoginViewModel: ObservableObject {

    // Inject AuthRepository from KMP shared module
    private let authRepository: AuthRepository = KoinHelper().authRepository

    // State yang akan diamati oleh SwiftUI
    @Published private(set) var loginState: LoginState = .idle

//    init(authRepository: AuthRepository) {
//        self.authRepository = authRepository
//    }

    // Fungsi untuk melakukan login
    func login(identifier: String, password: String) {
        loginState = .loading

        Task {
            do {
                // Panggil fungsi loginUser dari repository KMP
                let resource = try await authRepository.loginUser(
                    identifier: identifier,
                    password: password
                )

                // Konversi Resource dari Kotlin ke state Swift
                switch resource {
                case is ResourceSuccess<AuthResponse>:
                    let successResource = resource as! ResourceSuccess<AuthResponse>
                    self.loginState = .success(successResource.data)
                case let errorResource as ResourceError<AuthResponse>:
                    self.loginState = .error(errorResource.message)
                default:
                    // Fallback untuk state yang tidak terduga
                    self.loginState = .error("An unknown error occurred.")
                }
            } catch {
                self.loginState = .error(
                    "Failed to login. Please check your credentials."
                )
            }
        }
    }

}

File: composeApp/src/commonMain/kotlin/com/talangraga/talangragaumrohmobile/presentation/login/LoginViewModel.kt
package com.talangraga.talangragaumrohmobile.presentation.login

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.talangraga.talangragaumrohmobile.data.network.ApiResponse
import com.talangraga.talangragaumrohmobile.domain.repository.AuthRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach

class LoginViewModel(
    private val authRepository: AuthRepository
): ViewModel() {

    private val _loginState = MutableStateFlow<LoginState>(LoginState.Idle)
    val loginState: StateFlow<LoginState> = _loginState.asStateFlow()

    fun login(identifier: String, password: String) {
        _loginState.value = LoginState.Loading
        authRepository.login(identifier, password)
            .onEach { response ->
                when (response) {
                    is ApiResponse.Error -> {
                        _loginState.value = LoginState.Error(response.error.error?.message ?: "Unknown error")
                    }
                    is ApiResponse.Success -> {
                        _loginState.value = LoginState.Success
                    }
                }
            }
            .launchIn(viewModelScope)
    }

}